Sources/Sentry/SentryHub.m

```diff
@@ -374,4 +377,46 @@ - (void)setUser:(SentryUser *_Nullable)user
     }
 }
 
+- (void)captureEnvelope:(SentryEnvelope *)envelope
+{
+    SentryClient *client = _client;
+    if (nil == client) {
+        return;
+    }
+
+    [client captureEnvelope:[self updateSessionState:envelope]];
+}
+
+- (SentryEnvelope *)updateSessionState:(SentryEnvelope *)envelope
+{
+    if ([self envelopeContainsEventWithErrorOrHigher:envelope.items]) {
+        SentrySession *currentSession = [self incrementSessionErrors];
+
+        if (nil != currentSession) {
+            // Create a new envelope with the session update
+            NSMutableArray<SentryEnvelopeItem *> *itemsToSend =
+                [[NSMutableArray alloc] initWithArray:envelope.items];
+            [itemsToSend addObject:[[SentryEnvelopeItem alloc] initWithSession:currentSession]];
+
+            return [[SentryEnvelope alloc] initWithHeader:envelope.header items:itemsToSend];
+        }
+    }
+
+    return envelope;
+}
+
+- (BOOL)envelopeContainsEventWithErrorOrHigher:(NSArray<SentryEnvelopeItem *> *)items
+{
+    for (SentryEnvelopeItem *item in items) {
+        if ([item.header.type isEqualToString:SentryEnvelopeItemTypeEvent]) {
+            SentryLevel level = [SentrySerialization levelFromData:item.data];
```

Comment:
H: by default, an SDK might not have a `level` at all because the backend infers `kSentryLevelError` automatically, we need to account for this use case.
if the level is not there, assume its `kSentryLevelError`

Reply:
All SDKs have level, they might not set it to the event because of the reason you pointed out.
Knowing that, we can add the level to the event on RN, Flutter etc so we know we're sending it.

Reply:
@bruno-garcia while I agree that setting the level would be the right approach, it'd require a change on all top-level SDKs eg RN, Flutter, and a synced release, defaulting error here would be just easier, for now, we could fix that when fully deserializing the event on Cocoa, wdyt?

Reply:
@philipphofmann it's alright. it's clear now